common/crypto/attestation-api/conversion/dcap-direct/b64attestation2b64collateral.cpp

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright 2024 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <algorithm>
+#include <string>
+#include <cstring>
+#include "types/types.h"
+#include "base64/base64.h"
+#include "sgx_ql_lib_common.h"
+#include "sgx_dcap_quoteverify.h"
+
+uint8_t* serialize_collateral(sgx_ql_qve_collateral_t* c)
+{
+        uint32_t collateral_size = sizeof(sgx_ql_qve_collateral_t) + c->pck_crl_issuer_chain_size + c->root_ca_crl_size + c->pck_crl_size + c->tcb_info_issuer_chain_size + c->tcb_info_size + c->qe_identity_issuer_chain_size + c->qe_identity_size;
+        fprintf(stderr, "calculated collateral size: %d\n", collateral_size);
+
+        uint8_t* p = (uint8_t*)malloc(collateral_size);
+        if(p == NULL) return NULL;
+        memset(p, '\0', collateral_size);
+
+        // NOTICE / WARNING: here we copy the entire data structure with "meaningless" pointers
+        // pointers will have to be adjusted at deserialization time
+        memcpy(p, (uint8_t*)c, sizeof(sgx_ql_qve_collateral_t));
+        memcpy(p+sizeof(sgx_ql_qve_collateral_t), c->pck_crl_issuer_chain, c->pck_crl_issuer_chain_size);
+        memcpy(p+sizeof(sgx_ql_qve_collateral_t)+c->pck_crl_issuer_chain_size, c->root_ca_crl, c->root_ca_crl_size);
+        memcpy(p+sizeof(sgx_ql_qve_collateral_t)+c->pck_crl_issuer_chain_size+c->root_ca_crl_size, c->pck_crl, c->pck_crl_size);
+        memcpy(p+sizeof(sgx_ql_qve_collateral_t)+c->pck_crl_issuer_chain_size+c->root_ca_crl_size+c->pck_crl_size, c->tcb_info_issuer_chain, c->tcb_info_issuer_chain_size);
+        memcpy(p+sizeof(sgx_ql_qve_collateral_t)+c->pck_crl_issuer_chain_size+c->root_ca_crl_size+c->pck_crl_size+c->tcb_info_issuer_chain_size, c->tcb_info, c->tcb_info_size);
+        memcpy(p+sizeof(sgx_ql_qve_collateral_t)+c->pck_crl_issuer_chain_size+c->root_ca_crl_size+c->pck_crl_size+c->tcb_info_issuer_chain_size+c->tcb_info_size, c->qe_identity_issuer_chain, c->qe_identity_issuer_chain_size);
+        memcpy(p+sizeof(sgx_ql_qve_collateral_t)+c->pck_crl_issuer_chain_size+c->root_ca_crl_size+c->pck_crl_size+c->tcb_info_issuer_chain_size+c->tcb_info_size+c->qe_identity_issuer_chain_size, c->qe_identity, c->qe_identity_size);
```

Comment:
Hmm, i'm confused why refactoring our (de)serialization code into a single place (separate package) and using the less brittle pattern `size=sizof(nextfield); memcpy(moving_ptr, next_field, size); moving_ptr+=size;` has anything to do with what the dcap library provides? (of course, it would have been nice if dcap library would already have provided (de)serialization but that's besides my point?)

Reply:
I think the update addresses your comment